src/inputs.rs

```diff
@@ -520,16 +550,109 @@ impl StatelessSource {
         match self.0.call_method0(py, "next") {
             Err(stop_ex) if stop_ex.is_instance_of::<PyStopIteration>(py) => Ok(None),
             Err(err) => Err(err),
-            Ok(items) => {
-                Ok(Some(items.extract(py).reraise(
-                    "`next` method in StatelessSource did not return a list",
-                )?))
-            }
+            Ok(items) => Ok(Some(
+                items
+                    .extract(py)
+                    .reraise("`StatelessSource.next` did not return a list")?,
+            )),
         }
     }
 
+    fn next_awake(&self, py: Python) -> PyResult<Option<DateTime<Utc>>> {
+        self.0
+            .call_method0(py, "next_awake")
+            .reraise("error calling `StatelessSource.next_awake`")?
+            .extract(py)
+            .reraise("error converting `StatelessSource.next_awake` return value to datetime")
+    }
+
     fn close(self, py: Python) -> PyResult<()> {
         let _ = self.0.call_method0(py, "close")?;
         Ok(())
     }
 }
+
+/// Update `cooldown` depending on its current value and next_awake's value,
+/// with a fixed `now` moment.
+fn update_cooldown(
+    cooldown: &mut Option<chrono::Duration>,
+    next_awake: Option<chrono::DateTime<Utc>>,
+    now: chrono::DateTime<Utc>,
+) {
+    // Calculate the cooldown from next_awake and now.
+    let next_cooldown = next_awake.map(|na| na.signed_duration_since(now));
+
+    // Update cooldown, that will always be `Some` after this.
+    *cooldown = Some(match (&cooldown, next_cooldown) {
+        // If next_awake is None, just set the cooldown to zero
+        (_, None) => ZERO_DUR,
+        // If next_awake is Some, and there is no previous cooldown,
+        // set the cooldown to next_awake, capped at MIN_COOLDOWN
+        (None, Some(next_cooldown)) => next_cooldown.max(MIN_COOLDOWN),
+        // If next_awake is Some, and there is a previous cooldown,
+        // set the cooldown to the minimum between that and next_awake,
+        // capped at MIN_COOLDOWN
+        (Some(cooldown), Some(next_cooldown)) => next_cooldown.max(MIN_COOLDOWN).min(*cooldown),
+    })
+}
+
+/// Get the delay to wait for the next activation and reset `cooldown`
+pub fn extract_delay(cooldown: &mut Option<chrono::Duration>) -> Duration {
+    cooldown
+        // Use take to reset cooldown to None
+        .take()
+        // If the delay was never set, default to MIN_COOLDOWN
+        .unwrap_or(MIN_COOLDOWN)
+        // Convert to std::time::Duration.
+        .to_std()
+        // If this fails it means the cooldown ended
+        // up being a negative value, which is allowed in python's
+        // timedelta and in chrono::Duration, but not in std::time::Duration
+        .expect("to be able to convert cooldown to std::time::Duration")
```

Comment:
This was here because I capped the duration to a minimum of zero, so if the code ever crashed here it meant that that invariant changed. The comment was not really clear about that though. Anyway this is fixed with the changes from the other comment